### PR TITLE
fix(engine,#663): topologically sort cfg.Modules by dependsOn before RegisterModule

### LINKED ISSUE #663: Engine: BuildFromConfig ignores yaml-level `dependsOn:` for external-plugin modules (init-order race)
## Problem

`StdEngine.BuildFromConfig` (engine.go:381) walks `cfg.Modules` in slice order (engine.go:489) and registers each module with `app.RegisterModule(mod)` (engine.go:517). The subsequent `app.Init()` (engine.go:520) then walks modules in **registration order** for the external-plugin path — there is no dependency-aware topological sort.

For external plugins that declare `dependsOn:` keys in their YAML module config, the engine validates the keys but does not honor them at init time. This means a child module's `Init()` can fire before its parent's `Init()`, breaking any plugin that uses `Init()` to register runtime state (broker registries, factory tables, etc.) that downstream modules look up.

## Concrete impact: workflow-plugin-eventbus + BMW

BMW (buymywishlist) just shipped PR #279 switching its 6 consumers from NATS to pgchannel:
- `eventbus.broker` module → registers a runtime broker via `Init()` (calls `RegisterBrokerInstance`)
- `eventbus.stream` module → calls `RegisterStream` in `Init()`
- 6 × `eventbus.consumer` modules → call `RegisterConsumer` in `Init()`, look up the broker via `LookupRuntime(brokerRef)`

With canonical names (`bmw-eventbus`, `bmw-stream`, `bmw-consumer-*`), alphabetical iteration fires consumers before the broker, and `RegisterConsumer` fails with `broker not registered within 10s`.

**Current workaround in BMW** (GoCodeAlone/buymywishlist#279, commit `765c1c6`): rename broker → `aaa-bmw-eventbus`, stream → `aab-bmw-stream` so they sort lexicographically before the consumers. `dependsOn:` keys are kept on the consumer modules as documentation, but they are dead weight at engine init time.

This is a brittle workaround — adding any future module whose name sorts before `aaa-` would re-break the deploy.

## Reproduction (minimal)

```yaml
modules:
  - name: child
    type: eventbus.consumer  # any module whose Init() needs a parent
    config: {stream_name: x, consumer_name: y, broker_ref: parent}
    dependsOn: [parent]      # ignored
  - name: parent
    type: eventbus.broker
    config: {provider: pgchannel, broker_target: in_process, dsn: ...}
```

`BuildFromConfig` registers `child` then `parent`; `app.Init()` calls `child.Init()` first; `child` looks up `parent` in the broker registry → fails.

## What we'd like

Either:

**(a)** Engine-level: topological sort `cfg.Modules` by `dependsOn:` before `app.RegisterModule` so registration order respects declared deps. (Probably cheapest; the YAML already carries the data.)

**(b)** Modular-level: have `app.Init()` honor a `DependencyAware`-like interface that external-plugin module factories can implement (return `Dependencies() []string`) and modular sorts at Init time.

(a) seems strictly better for external plugins since the dependency is a *config-level* property of how the operator wired the modules, not a property of the module struct.

## Tag for the fix

When the engine-side fix lands, BMW will revert the `aaa-`/`aab-` prefixes in app.yaml.

## Related

- BMW PR: https://github.com/GoCodeAlone/buymywishlist/pull/279 (merged)
- Plugin: GoCodeAlone/workflow-plugin-eventbus v0.3.4

### PR DESCRIPTION
**Pull request by intel352**
## Summary

Closes #663 — engine init-order race for external-plugin modules.

`StdEngine.BuildFromConfig` walked `cfg.Modules` in slice order and registered each module via `app.RegisterModule`. The modular framework's `app.Init()` then walked modules in registration order — and for external-plugin module wrappers (which do not implement `DependencyAware`), yaml-level `dependsOn:` was validated by the schema but ignored at init time.

A child module whose `Init()` looked up runtime state in a plugin-local registry (broker, factory table, etc.) could fire before its parent's `Init()` had populated that state. **BMW PR 279 (eventbus pgchannel) hit this** as the `broker bmw-eventbus not registered within 10s` startup race and worked around it by renaming the broker module to `aaa-bmw-eventbus` so it sorted alphabetically before the consumers (commit `765c1c6` in that PR).

## Changes

- `engine_module_order.go`: new `topoSortModules` function. Kahn's algorithm with stable tie-break on declared index.
  - **Missing-dep targets tolerated** as no-op edges. `schema.ValidateConfig` already rejects those with an actionable error at `schema/validate.go:191`.
  - **Cycles return a descriptive error** listing the modules involved (sorted for deterministic output).
  - **Siblings keep declared order** — operators who hand-craft init order for non-dep reasons see no behavior change.
- `engine.go`: call `topoSortModules` once before the registration loop. Error surfaces with `module dependency ordering failed:` prefix.
- `CHANGELOG.md`: Fixed entry under [Unreleased].

## Tests

10 unit tests in `engine_module_order_test.go`:

- `TestTopoSortModules_NoDeps_PreservesOrder` — declared order preserved when no dependencies.
- `TestTopoSortModules_SingleChain` — C→B→A in input becomes A,B,C output.
- `TestTopoSortModules_BMWShape` — the exact PR 279 shape (broker last in declared order, consumers first, stream depends on broker). Verifies broker→stream→consumers ordering.
- `TestTopoSortModules_ParallelChains` — two independent chains, stable tie-break by index.
- `TestTopoSortModules_AlreadyCorrect_IsNoOp` — already-sorted input passes through unchanged.
- `TestTopoSortModules_MissingDepIsTolerated` — `dependsOn` pointing at a non-existent module name is silently ignored (separate validator catches it).
- `TestTopoSortModules_DirectCycleDetected` — A→B, B→A returns an error mentioning both modules.
- `TestTopoSortModules_IndirectCycleDetected` — A→C→B→A returns an error listing all three.
- `TestTopoSortModules_EmptyAndSingleton` — nil and 1-element inputs handled.
- `TestTopoSortModules_EmptyDependencyStringIgnored` — `dependsOn: ["", "a"]` defensively handled (schema rejects "" upstream).

```
$ GOWORK=off go test -race -count=1 -short -timeout 240s .
ok  	github.com/GoCodeAlone/workflow	9.543s
```

## Follow-up

Once this lands and a release tag ships:

- BMW (`GoCodeAlone/buymywishlist`) will revert `aaa-bmw-eventbus → bmw-eventbus` + `aab-bmw-stream → bmw-stream` in `app.yaml`, sweeping 13 references (broker_ref + dependsOn + trigger bus). Tracked as memory entry for autonomous follow-up.

## Test plan

- [x] All 10 new unit tests pass.
- [x] Engine package full suite passes with `-race`.
- [x] Config + schema package suites pass (no changes there, but DependsOn paths exercised).
- [ ] CI on this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: engine.go [hunk 1]
```diff
@@ -496,6 +496,22 @@ func (e *StdEngine) BuildFromConfig(cfg *config.WorkflowConfig) error {
...
+	// Reorder cfg.Modules so each module's RegisterModule call follows every
+	// module it lists in DependsOn. Resolves the init-order race that bit
+	// external-plugin consumers in BMW PR #279 (workflow#663): the modular
+	// app.Init() pass walks modules in registration order, and external-plugin
+	// module factories do not implement DependencyAware, so without this sort
+	// a consumer module's Init() can call into a broker registry that its
+	// dependency module has not populated yet. Schema.ValidateConfig has
+	// already verified that every dependsOn target exists and that there are
+	// no empty-string entries; topoSortModules tolerates a missing target as
+	// a defensive no-op edge.
```
[the remark this answers is not in the document]
Comment (intel352) [reply engine.go]: Addressed in cc42b3ee: dropped the misleading comment claim that schema.ValidateConfig has already validated the dependsOn references. Validation runs BEFORE ConfigTransformHooks, and transforms can inject modules whose deps were never schema-validated. topoSortModules is now documented as defensively tolerating a missing dependency target — schema rejects unresolvable deps for declared modules, and a transform-injected module with an unresolvable dep gets placed as if the edge didn't exist (the failure, if any, surfaces when the dependent looks up its parent at runtime).

File: engine_module_order.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+// orders the modules that *are* declared.
+func topoSortModules(modules []config.ModuleConfig) ([]config.ModuleConfig, error) {
+	n := len(modules)
+	if n <= 1 {
+		return modules, nil
+	}
+
+	index := make(map[string]int, n)
+	for i, m := range modules {
+		index[m.Name] = i
```
[the remark this answers is not in the document]
Comment (intel352) [reply engine_module_order.go]: Addressed in cc42b3ee: `index[m.Name] = i` now uses a first-wins guard (`if _, exists := index[m.Name]; !exists`) so duplicates resolve dependents against the original occurrence. Matches schema's first-occurrence semantic. Added TestTopoSortModules_DuplicateNames_FirstWins covering the behaviour.

File: engine.go [hunk 3]
```diff
@@ -496,6 +496,22 @@ func (e *StdEngine) BuildFromConfig(cfg *config.WorkflowConfig) error {
 		e.configHash = fmt.Sprintf("sha256:%x", h)
 	}
 
+	// Reorder cfg.Modules so each module's RegisterModule call follows every
+	// module it lists in DependsOn. Resolves the init-order race that bit
```
[the remark this answers is not in the document]
Comment (intel352) [reply engine.go]: Addressed in cc42b3ee: topoSortModules now runs BEFORE the configHash marshal+hash compute, so the stored hash reflects the order the engine actually builds from. The new ordering is: ConfigTransformHooks → topoSortModules → configHash compute → RegisterModule loop.

File: engine_module_order.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	"sort"
+	"strings"
+
+	"github.com/GoCodeAlone/workflow/config"
+)
+
+// topoSortModules reorders modules so each module appears after every module it
+// lists in DependsOn. Sibling modules (modules with no inter-dependency) keep
+// their declared order — Kahn's algorithm with a stable tie-break on the
+// original index. Returns an error if a dependency cycle is detected.
```
[the remark this answers is not in the document]
Comment (intel352) [reply engine_module_order.go]: Addressed in 2f3fd2db: docstring rewritten to state the actual invariant — "among modules in the ready frontier at any point, the one with the lowest declared index is dequeued first." Drops the misleading "siblings keep declared order" claim that didn't survive cross-chain pops.

File: CHANGELOG.md [hunk 5]
```diff
@@ -23,6 +23,10 @@ Configs that still reference the legacy types now fail to load with an actionabl
...
+
+- **`StdEngine.BuildFromConfig` now topologically sorts `cfg.Modules` by yaml `dependsOn:` before module registration.** Previously the engine walked `cfg.Modules` in slice order and the modular framework's `app.Init()` then walked in registration order. For external-plugin modules (which do not implement `DependencyAware`), yaml-level `dependsOn:` was validated by the schema but ignored at init time — a child module's `Init()` could fire before its parent's `Init()` had registered runtime state in a plugin-local registry (broker, factory table, etc.), causing startup races like the "broker not registered within 10s" failure that bit BMW PR #279. The new `topoSortModules` (in `engine_module_order.go`) uses Kahn's algorithm with a stable tie-break on declared index, tolerates missing dependency targets (schema validation catches those separately), and returns an error naming the cycle members if dependencies form a loop. 10 unit tests cover the BMW shape, parallel chains, cycles, and edge cases. Operators who used the alphabetical-prefix workaround (renaming a dependency to `aaa-…`) can now revert to canonical names and rely on `dependsOn:` to drive the order.
```
[the remark this answers is not in the document]
Comment (intel352) [reply CHANGELOG.md]: Addressed in 2f3fd2db: error message now reads "module dependsOn cycle (or dependents of one) among: <list>" and CHANGELOG matches. Added TestTopoSortModules_CycleErrorListsDependentsToo pinning the behaviour — explicitly verifies that a downstream dependent of a 2-cycle appears in the error message alongside the cycle members. Tightening to true SCC detection (Tarjan) is a separate refactor; the test will catch it as a behaviour change.

File: engine_module_order.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+// a module's final position can shift relative to its declared siblings if a
+// sibling becomes ready earlier (e.g., a root pops first and unblocks its
+// dependent, which now sits in the ready queue alongside other declared-later
+// roots). What is preserved is the relative order of modules that are *both*
+// in the ready frontier at the same iteration. Returns an error if a
+// dependency cycle is detected.
+//
+// Missing dependency targets are tolerated here and reported via the existing
+// schema.ValidateConfig pass (schema/validate.go:191); this function only
+// orders the modules that *are* declared.
```
[the remark this answers is not in the document]
Comment (intel352) [reply engine_module_order.go]: Addressed in 67559397: docstring now states the actual contract — missing dependency targets are silently tolerated as no-op edges, since ConfigTransformHooks can inject modules after schema.ValidateConfig has run and those injected modules' dependsOn was never validated. The schema reference is dropped.

File: engine_module_order.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+			ready = append(ready, i)
+		}
+	}
+
+	out := make([]config.ModuleConfig, 0, n)
+	for len(ready) > 0 {
+		// Stable tie-break: pop the lowest declared index first.
+		sort.Ints(ready)
+		i := ready[0]
+		ready = ready[1:]
```
[the remark this answers is not in the document]
Comment (intel352) [reply engine_module_order.go]: Addressed in 67559397: frontier is now a container/heap min-heap of declared indices. Push/Pop are O(log n) each, total O((V+E) log V). The old sort.Ints(ready) every-iteration call was O(n² log n) worst-case. Concrete impact is microseconds at BMW scale (~50 modules) either way, but the heap version is the obvious shape and removes the latent O(n²) tied to config size growing.
